pkg/ports/publish.go

```diff
@@ -216,23 +216,43 @@ func collectPorts(seen map[int32][]int32, seenHostnames map[string]struct{}, por
 			port.Port = port.TargetPort
 		}
 
-		if targets, ok := seen[port.Port]; ok {
-			// Check for special case: the same port is exposed on multiple hostnames, so keep both.
-			if port.Hostname != "" {
-				for _, t := range targets {
-					if t == port.TargetPort {
-						// Same port and target port but different hostnames, so keep both
-						seen[port.Port] = append(targets, port.TargetPort)
-						seenHostnames[port.Hostname] = struct{}{}
-						result = append(result, port)
+		protocolMatch := func(a, b v1.Protocol) bool {
+			return a == b ||
+				(a == v1.ProtocolHTTP && b == v1.ProtocolTCP) ||
+				(a == v1.ProtocolTCP && b == v1.ProtocolHTTP)
+		}
```

Comment:
This is something I don't like and I think we should discuss.
I feel like we should be smarter here.
I.e. HTTP should supersede TCP, as both create a TCP service, right? HTTP just adds an Ingress (if publish is true).
However, both should be able to co-exist if both a `publish: true`.
Any opinions?

Reply:
What about http2?  Or the next protocol that gets added?

Asking for multiple variants of tcp doesn't really make sense; e.g. tcp means the data within is opaque and untouched, http is changing it to add headers.  

The user should pick which one they want and each acorn port definition corresponds directly to exactly one layer-4 port in the container. 

Since we have various kinds of TCP, one UDP and no others, that check could currently be just "a is UDP and b is not".

Reply:
I agree that it doesn't make sense, but even http/tcp could be fine. Both result in a k8s service with tcp defined as protocol.
But: http *may* add Ingress. TCP *may* set type LoadBalancer. Http2 will create a separate k8s service with h2c annotation set.

I'm fine with being restrictive/opinionated here though just as you suggest 👍

Reply:
Yeah I'm not sure what I really think here either. In the issue that this PR links to, Bill has an example Acornfile provided by a user and they are defining TCP and UDP on the same port number. Is that something we want to support, and is it even possible?

Reply:
I guess at least for plain old tcp/udp via the same port we should have support, since that's used by DNS, RTP services, even OpenVPN or Minecraft Server.
That obviously depends on the configuration, but I think we should support it in general.
Not sure about all the other specific protocols though 🤷‍♂️

Reply:
I feel like multiple combinations are possible but potentially unlikely to be used at all.
For now I'll roll with Vince's suggestion of only allowing UDP + 1 flavor of TCP (tcp/http/http2) in general.